packages/components/src/calendar/test/index-range.tsx

```diff
@@ -0,0 +1,1212 @@
+/**
+ * @jest-environment jsdom
+ */
+
+import { render, screen, within } from '@testing-library/react';
+import userEvent from '@testing-library/user-event';
+import {
+	startOfDay,
+	startOfWeek,
+	startOfMonth,
+	endOfWeek,
+	addDays,
+	subDays,
+	addWeeks,
+	addMonths,
+	subMonths,
+	subYears,
+	addHours,
+} from 'date-fns';
+import { ar } from 'date-fns/locale';
+import { useState, default as React } from 'react';
+import '@testing-library/jest-dom';
+import { DateRangeCalendar, TZDate } from '..';
+import { getDateButton, getDateCell, monthNameFormatter } from '../utils/test-utils';
+import type { DateRange, DateRangeCalendarProps } from '../types';
+
+const UncontrolledDateRangeCalendar = (
+	props: DateRangeCalendarProps & {
+		initialSelected?: DateRange | undefined | null;
+		initialMonth?: Date | undefined;
+	}
+) => {
+	return (
+		<DateRangeCalendar
+			{ ...props }
+			defaultSelected={ props.initialSelected ?? undefined }
+			defaultMonth={ props.initialMonth }
+		/>
+	);
+};
+
+const ControlledDateRangeCalendar = (
+	props: DateRangeCalendarProps & {
+		initialSelected?: DateRange | undefined | null;
+		initialMonth?: Date | undefined;
+	}
+) => {
+	const [ selected, setSelected ] = useState< DateRange | undefined | null >(
+		props.initialSelected
+	);
+	const [ month, setMonth ] = useState< Date | undefined >( props.initialMonth );
+	return (
+		<DateRangeCalendar
+			{ ...props }
+			selected={ selected ?? null }
+			onSelect={ ( ...args ) => {
+				setSelected( args[ 0 ] );
+				props.onSelect?.( ...args );
+			} }
+			month={ month }
+			onMonthChange={ ( month ) => {
+				setMonth( month );
+				props.onMonthChange?.( month );
+			} }
+		/>
+	);
+};
+
+function setupUserEvent() {
+	// The `advanceTimersByTime` is needed since we're using jest
+	// fake timers to simulate a fixed date for tests.
+	return userEvent.setup( { advanceTimers: jest.advanceTimersByTime } );
+}
```

Comment:
Could we simplify this to assign a shared variable, rather than having to set this up in each test?

```suggestion
// import baseUserEvent from '@testing-library/user-event';
// ...

// The `advanceTimersByTime` is needed since we're using jest
// fake timers to simulate a fixed date for tests.
const userEvent = baseUserEvent.setup( { advanceTimers: jest.advanceTimersByTime } );
```

Reply:
We used to do that, but then found out in the [official docs](https://testing-library.com/docs/user-event/intro/#writing-tests-with-userevent) that it is recommended to run the `setup` inside each test before rendering the component being tested 🤷

Reply:
Interesting, thanks for sharing. The [blog post it references](https://kentcdodds.com/blog/avoid-nesting-when-youre-testing) makes some good points, though I think the issues it describes about indirection of assignment and potential for reassignment through nesting don't feel quite as relevant for this case, which is more of a one-time test setup incidental to our use of Jest timers in tests.

But not an opinion I hold too strongly, particularly if it's the recommendation of Testing Library.